File: framework/cei_python3/impl/restfulrequest.py
import urllib.parse
from impl.restfuloptions import RestfulOptions


class RestfulRequest:
    class Method:
        GET = "get"
        POST = "post"

    def __init__(self, option: RestfulOptions):
        self.__method = None
        self.__url = option.url
        self.__target = ""
        self.__post_body = ""
        self.__header = dict()
        self.__queryString = list()

    def build_query_string(self):
        if len(self.__queryString) == 0:
            return ""
        tmp = dict()
        for item in self.__queryString:
            tmp[item[0]]= item[1]
        encoded_param = urllib.parse.urlencode(tmp)
        return "?" + encoded_param

    def get_header(self):
        return self.__header

    def get_url(self):
        return self.__url

    def get_target(self):
        return self.__target

    def get_method(self):
        return self.__method

    def set_url(self, url: str):
        self.__url = url

    def set_target(self, target: str):
        self.__target = target

    def set_method(self, method: Method):
        self.__method = method

    def add_query_string(self, key: str, value):
        if value is not None:
            self.__queryString.append([key, value])

    def get_query_string(self):
        return self.__queryString

    def add_header(self, key: str, value):
        if value is None:
            # TODO
            pass
        self.__header[key] = str(value)

File: framework/cei_golang/src/ceiimpl/restfulrequest.go
package ceiimpl

import (
	"ceiimpl/utils"
)

type Method string

const (
	GET  Method = "GET"
	POST Method = "POST"
)

type RestfulRequest struct {
	url          string
	target       string
	method       Method
	headers      map[string]string
	queryStrings [][]string
	requestBody  []byte
}

func NewRestfulRequest(options *RestfulOptions) *RestfulRequest {
	instance := new(RestfulRequest)
	return instance
}

func (inst *RestfulRequest) SetTarget(target string, args ...interface{}) {
	inst.target, _ = utils.FillString(target, args)
}

func (inst *RestfulRequest) SetPostBody(json *JsonWrapper) {
	inst.requestBody = json.ToBytes()
}

func (inst *RestfulRequest) GetTarget() string {
	return inst.target
}

func (inst *RestfulRequest) SetMethod(method Method) {
	inst.method = method
}

func (inst *RestfulRequest) GetMethod() Method {
	return inst.method
}

func (inst *RestfulRequest) GetRequestBody() []byte {
	return inst.requestBody
}

func (inst *RestfulRequest) GetQueryString() [][]string {
	return inst.queryStrings
}

func (inst *RestfulRequest) AddHeader(key string, value interface{}) {
	inst.headers[key] = utils.ToString(value)
}

func (inst *RestfulRequest) AddQueryString(key string, value interface{}) {
	item := []string{key, utils.ToString(value)}
	inst.queryStrings = append(inst.queryStrings, item)
}

func (inst *RestfulRequest) GetHeaders() map[string]string {
	return inst.headers
}

func (inst *RestfulRequest) SetURL(value string) {
	inst.url = value
}

func (inst *RestfulRequest) GetURL() string {
	return inst.url
}

func (inst *RestfulRequest) BuildQueryString() string {
	result := ""
	for _, value := range inst.queryStrings {
		if result != "" {
			result += "&"
		}
		result += value[0]
		result += "="
		result += value[1]
	}
	return "?" + result
}
